File: python/optionslib/options_pricing.py
"""
Options Pricing Models Module.
Implements various options pricing models including Black-Scholes, Binomial, and Monte Carlo.
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from dataclasses import dataclass
from typing import Dict, Tuple, Optional
try:
    from optionslib_cpp import (
        OptionParamsCpp, OptionResultCpp,
        black_scholes_cpp, binomial_tree_cpp, monte_carlo_cpp
    )
    CPP_AVAILABLE = True
except ImportError:
    CPP_AVAILABLE = False
    print("Warning: C++ implementation not available. Using pure Python implementation.")

@dataclass
class OptionParams:
    """Parameters for option pricing."""
    S: float  # Current stock price
    K: float  # Strike price
    T: float  # Time to maturity
    r: float  # Risk-free rate
    sigma: float  # Volatility
    q: float = 0.0  # Dividend yield
    option_type: str = 'call'  # 'call' or 'put'

    def to_cpp_params(self) -> Optional['OptionParamsCpp']:
        if not CPP_AVAILABLE:
            return None
        params = OptionParamsCpp()
        params.S = self.S
        params.K = self.K
        params.T = self.T
        params.r = self.r
        params.sigma = self.sigma
        params.q = self.q
        params.is_call = self.option_type == 'call'
        return params

class OptionsPricing:
    """Options pricing models implementation."""
    
    @staticmethod
    def black_scholes(params: OptionParams) -> Dict[str, float]:
        """Calculate option price and Greeks using Black-Scholes model."""
        if CPP_AVAILABLE:
            cpp_params = params.to_cpp_params()
            result = black_scholes_cpp(cpp_params)
            return {
                'price': result.price,
                'delta': result.delta,
                'gamma': result.gamma,
                'theta': result.theta,
                'vega': result.vega,
                'rho': result.rho
            }
        
        # Pure Python implementation as fallback
        S, K, T, r, sigma, q = params.S, params.K, params.T, params.r, params.sigma, params.q
        is_call = params.option_type == 'call'
        
        # Calculate d1 and d2
        d1 = (np.log(S/K) + (r - q + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
        d2 = d1 - sigma * np.sqrt(T)
        
        # Calculate option price
        if is_call:
            price = S * np.exp(-q * T) * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
        else:
            price = K * np.exp(-r * T) * norm.cdf(-d2) - S * np.exp(-q * T) * norm.cdf(-d1)
        
        # Calculate Greeks
        delta = np.exp(-q * T) * (norm.cdf(d1) if is_call else -norm.cdf(-d1))
        gamma = np.exp(-q * T) * norm.pdf(d1) / (S * sigma * np.sqrt(T))
        theta = (-S * sigma * np.exp(-q * T) * norm.pdf(d1) / (2 * np.sqrt(T)) -
                r * K * np.exp(-r * T) * (norm.cdf(d2) if is_call else norm.cdf(-d2)) +
                q * S * np.exp(-q * T) * (norm.cdf(d1) if is_call else norm.cdf(-d1)))
        vega = S * np.exp(-q * T) * np.sqrt(T) * norm.pdf(d1)
        rho = K * T * np.exp(-r * T) * (norm.cdf(d2) if is_call else -norm.cdf(-d2))
        
        return {
            'price': price,
            'delta': delta,
            'gamma': gamma,
            'theta': theta,
            'vega': vega,
            'rho': rho
        }
    
    @staticmethod
    def binomial_tree(params: OptionParams, steps: int = 100) -> Dict[str, float]:
        """Calculate option price and Greeks using Binomial Tree model."""
        if CPP_AVAILABLE:
            cpp_params = params.to_cpp_params()
            result = binomial_tree_cpp(cpp_params, steps)
            return {
                'price': result.price,
                'delta': result.delta,
                'gamma': result.gamma,
                'theta': result.theta,
                'vega': result.vega,
                'rho': result.rho
            }
        
        # Pure Python implementation as fallback
        S, K, T, r, sigma, q = params.S, params.K, params.T, params.r, params.sigma, params.q
        is_call = params.option_type == 'call'
        
        # Calculate parameters
        dt = T / steps
        u = np.exp(sigma * np.sqrt(dt))
        d = 1 / u
        p = (np.exp((r - q) * dt) - d) / (u - d)
        
        # Initialize stock price tree
        stock = np.zeros((steps + 1, steps + 1))
        stock[0, 0] = S
        
        for i in range(1, steps + 1):
            stock[i, 0] = stock[i-1, 0] * u
            for j in range(1, i + 1):
                stock[i, j] = stock[i-1, j-1] * d
        
        # Initialize option value tree
        option = np.zeros((steps + 1, steps + 1))
        
        # Calculate option values at expiration
        for j in range(steps + 1):
            if is_call:
                option[steps, j] = max(0, stock[steps, j] - K)
            else:
                option[steps, j] = max(0, K - stock[steps, j])
        
        # Backward induction
        df = np.exp(-r * dt)
        for i in range(steps - 1, -1, -1):
            for j in range(i + 1):
                option[i, j] = df * (p * option[i+1, j] + (1-p) * option[i+1, j+1])
        
        price = option[0, 0]
        
        # Calculate Greeks using finite differences
        delta = (option[1, 0] - option[1, 1]) / (stock[1, 0] - stock[1, 1])
        gamma = ((option[2, 0] - option[2, 1]) / (stock[2, 0] - stock[2, 1]) -
                (option[2, 1] - option[2, 2]) / (stock[2, 1] - stock[2, 2])) / (0.5 * (stock[2, 0] - stock[2, 2]))
        
        return {
            'price': price,
            'delta': delta,
            'gamma': gamma
        }
    
    @staticmethod
    def monte_carlo(params: OptionParams, num_sims: int = 100000, time_steps: int = 100) -> Tuple[float, float]:
        """Calculate option price using Monte Carlo simulation."""
        if CPP_AVAILABLE:
            cpp_params = params.to_cpp_params()
            return monte_carlo_cpp(cpp_params, num_sims, time_steps)
        
        # Pure Python implementation as fallback
        S, K, T, r, sigma, q = params.S, params.K, params.T, params.r, params.sigma, params.q
        is_call = params.option_type == 'call'
        
        dt = T / time_steps
        nudt = (r - q - 0.5 * sigma**2) * dt
        sigsdt = sigma * np.sqrt(dt)
        
        # Generate paths
        Z = np.random.standard_normal((num_sims, time_steps))
        paths = np.zeros((num_sims, time_steps + 1))
        paths[:, 0] = S
        
        for t in range(time_steps):
            paths[:, t+1] = paths[:, t] * np.exp(nudt + sigsdt * Z[:, t])
        
        # Calculate payoffs
        if is_call:
            payoffs = np.maximum(paths[:, -1] - K, 0)
        else:
            payoffs = np.maximum(K - paths[:, -1], 0)
        
        # Calculate price
        price = np.exp(-r * T) * np.mean(payoffs)
        
        # Calculate standard error
        std_error = np.std(payoffs) / np.sqrt(num_sims)
        
        return price, std_error
    
    @staticmethod
    def plot_option_surface(spot_range: np.ndarray, vol_range: np.ndarray, params: OptionParams) -> None:
        """Plot option price surface as a function of spot price and volatility."""
        X, Y = np.meshgrid(spot_range, vol_range)
        Z = np.zeros_like(X)
        
        for i in range(len(vol_range)):
            for j in range(len(spot_range)):
                temp_params = OptionParams(
                    S=spot_range[j],
                    K=params.K,
                    T=params.T,
                    r=params.r,
                    sigma=vol_range[i],
                    q=params.q,
                    option_type=params.option_type
                )
                
                if CPP_AVAILABLE:
                    cpp_params = temp_params.to_cpp_params()
                    result = black_scholes_cpp(cpp_params)
                    Z[i, j] = result.price
                else:
                    result = OptionsPricing.black_scholes(temp_params)
                    Z[i, j] = result['price']
        
        fig = plt.figure(figsize=(12, 8))
        ax = fig.add_subplot(111, projection='3d')
        surface = ax.plot_surface(X, Y, Z, cmap='viridis')
        
        ax.set_xlabel('Spot Price')
        ax.set_ylabel('Volatility')
        ax.set_zlabel('Option Price')
        ax.set_title(f'{params.option_type.title()} Option Price Surface')
        
        plt.colorbar(surface)
    
    @staticmethod
    def plot_greeks(spot_range: np.ndarray, params: OptionParams) -> None:
        """Plot option Greeks as a function of spot price."""
        greeks = ['delta', 'gamma', 'theta', 'vega', 'rho']
        values = {greek: [] for greek in greeks}
        
        for spot in spot_range:
            temp_params = OptionParams(
                S=spot,
                K=params.K,
                T=params.T,
                r=params.r,
                sigma=params.sigma,
                q=params.q,
                option_type=params.option_type
            )
            if CPP_AVAILABLE:
                cpp_params = temp_params.to_cpp_params()
                result = black_scholes_cpp(cpp_params)
                for greek in greeks:
                    values[greek].append(getattr(result, greek))
            else:
                result = OptionsPricing.black_scholes(temp_params)
                for greek in greeks:
                    values[greek].append(result[greek])
        
        fig, axes = plt.subplots(2, 3, figsize=(15, 10))
        axes = axes.ravel()
        
        for idx, greek in enumerate(greeks):
            axes[idx].plot(spot_range, values[greek])
            axes[idx].set_xlabel('Spot Price')
            axes[idx].set_ylabel(greek.title())
            axes[idx].set_title(f'{greek.title()} vs Spot Price')
            axes[idx].grid(True)
        
        plt.tight_layout()

File: cpp/optionslib/options_pricing.cpp
#include "options_pricing.hpp"
#include <cmath>

namespace optionslib {

double OptionPricingCpp::norm_cdf(double x) {
    return 0.5 * (1 + std::erf(x / std::sqrt(2)));
}

double OptionPricingCpp::norm_pdf(double x) {
    return (1.0 / std::sqrt(2 * M_PI)) * std::exp(-0.5 * x * x);
}

OptionPricingCpp::OptionResult OptionPricingCpp::black_scholes(const OptionParams& params) {
    double d1 = (std::log(params.S / params.K) + (params.r - params.q + 0.5 * params.sigma * params.sigma) * params.T) / 
                (params.sigma * std::sqrt(params.T));
    double d2 = d1 - params.sigma * std::sqrt(params.T);
    
    OptionResult result;
    if (params.is_call) {
        result.price = params.S * std::exp(-params.q * params.T) * norm_cdf(d1) - 
                      params.K * std::exp(-params.r * params.T) * norm_cdf(d2);
        result.delta = std::exp(-params.q * params.T) * norm_cdf(d1);
    } else {
        result.price = params.K * std::exp(-params.r * params.T) * norm_cdf(-d2) - 
                      params.S * std::exp(-params.q * params.T) * norm_cdf(-d1);
        result.delta = -std::exp(-params.q * params.T) * norm_cdf(-d1);
    }
    
    result.gamma = std::exp(-params.q * params.T) * norm_pdf(d1) / 
                  (params.S * params.sigma * std::sqrt(params.T));
    result.vega = params.S * std::exp(-params.q * params.T) * std::sqrt(params.T) * norm_pdf(d1);
    result.theta = -params.S * std::exp(-params.q * params.T) * norm_pdf(d1) * params.sigma / 
                  (2 * std::sqrt(params.T)) - params.r * params.K * std::exp(-params.r * params.T) * norm_cdf(d2);
    result.rho = params.K * params.T * std::exp(-params.r * params.T) * norm_cdf(d2);
    
    return result;
}

OptionPricingCpp::OptionResult OptionPricingCpp::binomial_tree(const OptionParams& params, int steps) {
    double dt = params.T / steps;
    double u = std::exp(params.sigma * std::sqrt(dt));
    double d = 1.0 / u;
    double p = (std::exp((params.r - params.q) * dt) - d) / (u - d);
    
    std::vector<double> prices(steps + 1);
    for (int i = 0; i <= steps; ++i) {
        prices[i] = params.S * std::pow(u, steps - i) * std::pow(d, i);
    }
    
    std::vector<double> values(steps + 1);
    for (int i = 0; i <= steps; ++i) {
        values[i] = params.is_call ? std::max(0.0, prices[i] - params.K) 
                                 : std::max(0.0, params.K - prices[i]);
    }
    
    for (int j = steps - 1; j >= 0; --j) {
        for (int i = 0; i <= j; ++i) {
            values[i] = std::exp(-params.r * dt) * (p * values[i] + (1 - p) * values[i + 1]);
        }
    }
    
    OptionResult result;
    result.price = values[0];
    
    // Calculate delta using finite difference
    double delta_s = 0.01 * params.S;
    OptionParams up_params = params;
    up_params.S += delta_s;
    OptionParams down_params = params;
    down_params.S -= delta_s;
    
    double up_price = binomial_tree(up_params, steps).price;
    double down_price = binomial_tree(down_params, steps).price;
    result.delta = (up_price - down_price) / (2 * delta_s);
    result.gamma = (up_price - 2 * result.price + down_price) / (delta_s * delta_s);
    
    // Set other Greeks to 0 for now
    result.theta = 0;
    result.vega = 0;
    result.rho = 0;
    
    return result;
}

std::pair<double, double> OptionPricingCpp::monte_carlo(const OptionParams& params, 
                                                       int num_sims, 
                                                       int time_steps) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::normal_distribution<> normal(0, 1);
    
    double dt = params.T / time_steps;
    double drift = (params.r - params.q - 0.5 * params.sigma * params.sigma) * dt;
    double vol = params.sigma * std::sqrt(dt);
    double discount = std::exp(-params.r * params.T);
    
    std::vector<double> payoffs(num_sims);
    for (int i = 0; i < num_sims; ++i) {
        double S_t = params.S;
        for (int t = 0; t < time_steps; ++t) {
            S_t *= std::exp(drift + vol * normal(gen));
        }
        
        payoffs[i] = params.is_call ? std::max(0.0, S_t - params.K) 
                                  : std::max(0.0, params.K - S_t);
    }
    
    double sum = 0.0;
    double sum_sq = 0.0;
    for (double payoff : payoffs) {
        sum += payoff;
        sum_sq += payoff * payoff;
    }
    
    double price = discount * sum / num_sims;
    double variance = (sum_sq - sum * sum / num_sims) / (num_sims - 1);
    double std_error = std::sqrt(variance / num_sims);
    
    return {price, std_error};
}

} // namespace optionslib
